IEEE Xplore title em dash cleanup

InAnYan opened the pull request:
<!-- 
Describe the changes you have made here: what, why, ... 
Link the issue that will be closed, e.g., "Closes #333".
If your PR closes a koppor issue, link it using its URL, e.g., "Closes https://github.com/koppor/jabref/issues/47".
"Closes" is a keyword GitHub uses to link PRs with issues; do not change it.
Don't reference an issue in the PR title because GitHub does not support auto-linking there.
-->
Closes https://github.com/koppor/jabref/issues/286.

I added a test to DoiFetcherTest for doi [https://doi.org/10.1109/PERCOMW.2015.7133989](https://doi.org/10.1109/PERCOMW.2015.7133989). 

The parsed title of the article should be:
```
Towards situation-aware adaptive workflows: SitOPT — A general purpose situation-aware workflow management system
```

But previously it was:
```
Towards situation-aware adaptive workflows: SitOPT &amp;#x2014; A general purpose situation-aware workflow management system
```

In order to make the cleanup, I added some code to the `format` method of the `FieldContentFormatter`. (Is that the right file for cleanup?).

### Mandatory checks
- [x] Change in CHANGELOG.md described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

InAnYan commented:
Oops, sorry. The test should be in field format test, but not in `DoiFetchTester`. Moved

Siedlerchr commented on src/main/java/org/jabref/logic/bibtex/FieldContentFormatter.java:
```diff
@@ -47,7 +47,7 @@ public String format(String fieldContent, Field field) {
         }
 
-        return WHITESPACE.matcher(fieldContent).replaceAll(" ");
+        return WHITESPACE.matcher(fieldContent).replaceAll(" ").replaceAll("&amp;#x2014;", "—");
```
This is not the right place as  this can have unforseen consequences

better create a new formatter, see for example NormalizeEnDashesFormatter

InAnYan replied:
Okay, if I create new formatter how to add it to the parser? Or just where to add it, so it will format strings?

InAnYan commented:
I've created a new formatter, but can't understand where to add it to format fields. Currently, the title for article [https://doi.org/10.1109/PERCOMW.2015.7133989](https://doi.org/10.1109/PERCOMW.2015.7133989) is not formatted. But I've made the class and tests. Maybe I should've not added it to the `Formatters` class (`NormalizeEnDashesFormatter` is not added there too)

Siedlerchr commented:
There should be a method doPostcleaup where you can add the formatter
If you want to apply for all fields, you can use InternalField. INTERNAL_ALL_FIELD

koppor commented on src/main/java/org/jabref/logic/formatter/bibtexfields/NormalizeEmDashesIEEEFormatter.java:
```diff
@@ -0,0 +1,31 @@
+import org.jabref.logic.cleanup.Formatter;
+import org.jabref.logic.l10n.Localization;
+
+public class NormalizeEmDashesIEEEFormatter extends Formatter {
```
Remove IEEE. Also from the description. The used character is a normal unicode character. Not special for IEEE.


Related formatter: https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/formatter/bibtexfields/NormalizeEnDashesFormatter.java

koppor commented on src/main/java/org/jabref/logic/formatter/bibtexfields/NormalizeEmDashesIEEEFormatter.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public String format(String value) {
+        return value.replaceAll("&amp;#x2014;", "—");
```
Seeing the input, isnt that a double encoded unicode character? 

Should the importer run https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/formatter/bibtexfields/HtmlToUnicodeFormatter.java twice for the title field?

InAnYan replied:
But that will leave the ampersand.

I've tried to add additional [HtmlToUnicodeFormatter.java](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/formatter/bibtexfields/HtmlToUnicodeFormatter.java) to `DoiFetcher.doPostCleanup`. Is that the right place? 

It replaced the `&amp;` with ampersand, but `#x2014;` was left as is. So, probably there should be a separate formatter?

I read the original issue and you have provided a sketch of the solution. You said to add a new cleanup class and integrate it with `IEEE` class. But how to add the cleanup to the `IEEE`? There seems to be no `doPostCleanup` method. What if we just change the `parseJsonResponse` and directly cleanup from there?

Siedlerchr replied:
For the IEEE doPostCleanup  method, it's  just a  single method that is called manually because only the EntryBasedParserFetcher has it defined as interface method.
See thee DoiFetcher for example, it defines it manually

InAnYan replied:
It seems that the JabRef doesn't use `IEEE` class when importing the DOI [https://doi.org/10.1109/PERCOMW.2015.7133989](https://doi.org/10.1109/PERCOMW.2015.7133989)

koppor reviewed (changes requested):
I think, running https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/formatter/bibtexfields/HtmlToUnicodeFormatter.java twice should be tested instead of introducing a new cleaning

InAnYan commented:
I've changed the class for the formatter.

Calling `HtmlToUnicodeFormatter` twice doesn't solve the problem completely.

And it seems `IEEE` class is not used for fetching [https://doi.org/10.1109/PERCOMW.2015.7133989](https://doi.org/10.1109/PERCOMW.2015.7133989). (Maybe that's because this is an old issue, and the way JabRef works has changed?).

I've added the formatter to `DoiFetcher.postCleanup`. It works.

Siedlerchr commented:
Pleease the failing tests here 
<img width="894" alt="grafik" src="https://github.com/JabRef/jabref/assets/320228/76ee7570-c3f2-4773-b3ee-a8dd7afdd073">
just click on details
for openrewrite you can execute the `rewriteRun` task

koppor commented on src/main/java/org/jabref/logic/formatter/bibtexfields/NormalizeAmpersandEmDashesFormatter.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public String format(String value) {
+        return value.replace("&amp;#x2014;", "—");
```
This is Unicode or plain ascii? I think, we fetch in plain BibTeX, thus, it should be --, shouldn't it?

InAnYan replied:
The source code is in Unicode. 

In `NormalizeEnDashesFormatter` en dash was replaced with `--`, em dash should be replaced with it too or something longer?

koppor replied:
Let's keep `--`.

Normally, it would be `---`, but this is very strange in Paper titles. Never saw it. Therefore, I proposed `--`.

koppor commented:
> Calling `HtmlToUnicodeFormatter` twice doesn't solve the problem completely.

There needed to be a small fix. I coded it in https://github.com/JabRef/jabref/pull/11091. Since the new code is much more smaller than the code proposed in this PR (and also handles more cases), I close this PR. 

Thank you for digging out where to call the formatter! 👍
